### PR TITLE
Feature 1157 importer for spdx json data

### LINKED ISSUE #1153: Introduce `licenseScan` scan type
## Situation
This is a kind of sub issue of #1152 

## Wanted
We want a new scan type: `licenseScan`

## Solution
- Introduce new scantype in SecHub configuration model
- Update documentation
- We need #1098 implement before to have the data section usage possiblity

Format inside sechub configuration model JSON (as fictional suggested by #1098):
```json
{
  "apiVersion" : "1.0",
  "server" : "https://sechub.example.org",
  "project" : "gamechanger",
  "data" : {
    "sources" : [ {
      "name" : "android-sources",
      "fileSystem" : {
        "folders" : [ "gamechanger-android/src/main/java", "gamechanger-server/src/main/java" ]
      }
    }],
    "binaries" : [ {
      "name" : "android-libraries",
      "fileSystem" : {
        "folders" : [ "gamechanger-android/lib/dlls", "gamechanger-server/lib/so" ]
      }
    }]
  "licenseScan" : {
    "use" : [ "android-sources","android-libraries" ]
  }
}
```

### LINKED ISSUE #1157: Extend Sereco internal format + provide importer for SPDX json data
## Situation
This is a kind of sub issue of #1152 

With #1156  we want to provide SPDX json information. But currently there is no such information available.

## Wanted
- Sereco shall have an importer for SPDX json
- Sereco shall contain the SPDX json information

## Solution
- Implement an importer for SPDX json into Sereco reports
- The SPDX json shall be contained "as is", so easy to fetch
- The SPDX information is optional

### PR DESCRIPTION
**Pull request by Jeeppler**
Implement license scanning and SPDX JSON

- Importer for SPDX JSON data #1157
- Introduce a license scan type on Server and PDS side #1153
- New REST interface to download SPDX reports #1156

This PR
- closes #1153
- closes #1157 
- closes: #1156

### CODE REVIEW

File: sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SpdxV1JSONImporterTest.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        /* execute */
+        ProductImportAbility ableToImportScancodeSpdx = importerToTest.isAbleToImportForProduct(paramScancode);
+
+        /* test */
+        assertEquals(ProductImportAbility.ABLE_TO_IMPORT, ableToImportScancodeSpdx, "Was NOT able to import SPDX JSON!");
+    }
+
+    @Test
+    void createImportSupport__import_spdx_scancode_binaryscan() {
```
Comment (de-jcup) [inline sechub-sereco/src/test/java/com/mercedesbenz/sechub/sereco/importer/SpdxV1JSONImporterTest.java:71]: Same as my comment before-  I would drop this.

File: sechub-shared-kernel/src/test/java/com/mercedesbenz/sechub/sharedkernel/configuration/SecHubConfigurationTest.java [hunk 2]
```diff
@@ -522,4 +524,22 @@ public void webscan_empty_includes_excludes() {
         assertTrue("excludes are empty", excludes.get().isEmpty());
         assertEquals(expectedExcludes, excludes.get());
     }
+
+    @Test
+    public void license_scan_test() {
```
Comment (de-jcup) [inline sechub-shared-kernel/src/test/java/com/mercedesbenz/sechub/sharedkernel/configuration/SecHubConfigurationTest.java:529]: Please rename, so meaning of test becomes clearer:
```suggestion
    public void a_sechub_configuration_JSON_with_license_scan_can_be_read_and_license_scan_has_correct_data_configuration_reference) {
```

**Review by de-jcup — CHANGES_REQUESTED** [review]
Good job.

But there are some parts where some changes are necessary. 
1. See my comments
2. There is no documentation in any .adoc file about the new scan type: licenseScan
   candidates would be:
  - pds_config.adoc
  - sechub_config.adoc
  
  Either we create another isseue for the documentation parts or it must be handled here.

File: sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubScanConfigurationTest.java [hunk 3]
```diff
@@ -24,4 +26,68 @@ void sechub_job_config_example1_JSON_can_be_desrialized_and_contains_expected_lo
...
+        /* test */
+        assertFalse(scanConfig.getWebScan().isPresent());
+        assertFalse(scanConfig.getLicenseScan().isPresent());
+        assertFalse(scanConfig.getCodeScan().isPresent());
+        assertFalse(scanConfig.getInfraScan().isPresent());
+
+        assertEquals("1.0", scanConfig.getApiVersion());
+        assertEquals(1, scanConfig.getData().get().getSources().size());
+
+        SecHubSourceDataConfiguration sourceConfig = scanConfig.getData().get().getSources().iterator().next();
```
[the remark this answers is not in the document]
Comment (Jeeppler) [reply sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubScanConfigurationTest.java:74]: The test is intended to test that no scan type is detected, but it can be read. You understood my intention. I also, wanted to test, that even if no proper scan type is defined the data section is still populated. The purpose of both is to test edge cases which the user might get wrong.

File: sechub-commons-model/src/test/resources/sechub_license_scan_wrong_source_config_example.json [hunk 4]
```diff
@@ -0,0 +1,14 @@
+{
+  "apiVersion" : "1.0",
+  "data" : {
+    "source" : [ {
```
[the remark this answers is not in the document]
Comment (Jeeppler) [reply sechub-commons-model/src/test/resources/sechub_license_scan_wrong_source_config_example.json:4]: While testing I used "source" instead of "sources" it took me quite some time to understand where the issue is. I wrote the test to check what happens if I just write "source" instead of "sources". 

The question would be should we give a hint in the error message? For example, by writing: `"source" is not acceptable. Do you mean "sources"`.

File: sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/ProductResultRepository.java [hunk 5]
```diff
@@ -2,6 +2,7 @@
 package com.mercedesbenz.sechub.domain.scan.product;
 
 import static com.mercedesbenz.sechub.domain.scan.product.ProductResult.*;
+import static com.mercedesbenz.sechub.domain.scan.product.ProductResult.TABLE_NAME;
```
[the remark this answers is not in the document]
Comment (Jeeppler) [reply sechub-scan/src/main/java/com/mercedesbenz/sechub/domain/scan/product/ProductResultRepository.java:5]: I applied the `spotless` formatter, but it apparently did not organize the imports in this case.

File: sechub-commons-model/src/test/resources/sechub_license_scan_wrong_source_config_example.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jeeppler) [reply sechub-commons-model/src/test/resources/sechub_license_scan_wrong_source_config_example.json:4]: I will remove this test and instead open an issue for the client. The client can analyze the user provided configuration and give some hints.

**Review by de-jcup — APPROVED** [review]
Looks fine for me.
